internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -215,11 +216,21 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 
 	events := []event.Event{}
 
+	var env *v1beta1.Environment
+	if req.Environment != nil {
+		d, err := AsStruct(req.Environment)
+		if err != nil {
+			return CompositionResult{}, errors.Wrap(err, errEnvAsStruct)
+		}
+		env = &v1beta1.Environment{
+			Data: d,
+		}
+	}
```

Comment:
I feel pretty genuinely ambivalent about that, haha.

I do really like the idea of moving the "bootstrapping" of the environment into a Function. This gives us more room to experiment there, and allows folks to opt out of the complexity if they don't need it.

What I'm not so sure about is the general pattern of Functions reading stuff from the API server and passing them to the Function pipeline. (I presume a Function would select and read EnvConfigs). Is that something we should allow? Or do we want Crossplane to be the sole broker of everything read from and written to the API server. I'm not concretely sure of what risk Functions reading/writing to the API server poses, more of an intuition at this point.

If we did allow this we'd probably need to update the spec: https://github.com/crossplane/crossplane/blob/5797021/contributing/specifications/functions.md#runtime-environment

Reply:
Yes, adding side effects could make everything more complex, e.g. how do we handle that from xrender? However I feel it's going to be hard to prevent, people can always add a service account and use that to do whatever they want, so I guess we should just embrace it and make it a thing we take into account. For example, we could add an annotation to let xrender know a function should be skipped or add conditionals as we were discussing in some other thread, so that one can render compositions skipping the retrieval steps and passing stuff into the environment instead of letting it be retrieved from the outside world.

Reply:
> Yes, adding side effects could make everything more complex, e.g. how do we handle that from xrender? However I feel it's going to be hard to prevent, people can always add a service account and use that to do whatever they want, so I guess we should just embrace it and make it a thing we take into account. 

I am pretty skeptical to the idea of natively supporting direct access to Kube API from a function. This is mostly because, this would open a door invalidating most of the opinionated framework that we're trying to provide with `RunFunctionRequest`/`RunFunctionResponse`. If a function could access to the Kube API native, why should it bother getting resources from Observations or specifying the desired state instead of just getting/patching the resources directly. I am imagining that there will be a tendency to open that door with the first limitation in the framework that somebody hits and never get back to track even we resolve that afterwards. There is also a whole new complexity that would be introduced due to the RBAC. How and who will define the RBAC for such a service account, if it is users, then this will still require manual work and passing a service account wouldn't solve the problem. I don't want to think about Crossplane also taking care of RBAC for such a case 😅 

However, considering the Kube API is just yet another API and we are planning to enable passing credentials to functions, power users could still communicate with the Kube API if/once they really have to. 

This feels quite similar to what one can do with **provider-kubernetes** and **provider-helm**. They _can_ be used to create resources (other than the providers own CRDs and a limited set of other types) on the same Kubernetes cluster if a provider config provided accordingly. However, we didn't enable this as a native functionality for all Crossplane Providers, and working fine so far.

Reply:
Let me invalidate a bit what I wrote above 😄 

I just remembered there was a functionality in the Provider Meta definition that no provider uses currently (to the best of my knowledge), which is [PermissionRequests](https://github.com/crossplane/crossplane/blob/master/apis/pkg/meta/v1/provider_types.go#L38-L42). I am not sure if it got broken at some point but if we reflect that functionality to Function meta package, an environment-merger function should be able to give a list RBAC policy rules in its package definition to access EnvironmentConfig resources. And this doesn't feel that bad as it will be something defined at the package level and have a very limited scope.

Reply:
Yes, having it already implemented makes it feel less bad, isn't it? 😂

I think it makes sense to embrace it, nothing is going to stop a power user from doing that already, a function is a pod (atm), and it gets a serviceAccount, so we should just accept it and help making it a good and controlled experience.

I don't really agree with the argument that if you are talking directly to the API server, you should just do that instead of going through Compositions. That's not really how I see it. The one who is willing to get into the complexity necessary to talk with the api server is the function author, which is probably an expert and wants to expose some simplified interface to Compositions users. If we merge the two personas, Compositions themselves become useless, as people could "just build their own Operator".

Reply:
> Yes, having it already implemented makes it feel less bad, isn't it? 😂

Not really 🙂
For me, the key here is RBAC and how/where it is defined.
Initially, I was thinking it as if it would enable/promote doing anything on the Kube API and functions evolving to k8s operators.
However, a permission model where only an explicit/limited set of resources defined by function author is accessible  makes it more controlled as you are stating and not harmful to our opinionated framework

Reply:
If I understand this correctly it means that while Functions can be used to modify the current environment/context, and could populate the environment from non-K8S API sources, they should not populate it from K8S API resources and that functionality needs to stay where it is (for now) in the Composition, and the discussion in #4583 should stay out of the realm of Functions.

Reply:
@negz
> As a concrete example, a Function that read EnvConfigs from the API server would not work with xrender. Or rather, you'd need to somehow plumb it up to an API server to allow it to get access to the types it needs.

Same issue would apply to object retrieval defined at composition level, even more with what we are discussing in #4583.

Reply:
@phisco I don't think it would be quite the same. `xrender` would need to be aware that a Composition could reference EnvConfigs and have a flag like `--env-configs` or something to pass in the EnvConfigs, similar to the `--observed-resources` flag.

That all said, maybe there's a middle ground here where a Function could ask Crossplane to provide the pipeline with "extra" observed resources. I wrote up a sketch of how that might work in https://github.com/crossplane/crossplane/issues/4739.